src/devices/video/315_5124.cpp

```diff
@@ -878,47 +942,49 @@ void sega315_5124_device::draw_scanline_mode4( int *line_buffer, int *priority_s
 				pixel_plot_x = 7 - pixel_x;
 			}
 
-			pixel_plot_x = (0 - (x_scroll & 0x07) + (tile_column << 3) + pixel_plot_x);
-			if (pixel_plot_x >= 0 && pixel_plot_x < 256)
+			pixel_plot_x = x_scroll_fine_adjust + (tile_column << 3) + pixel_plot_x;
+			if (pixel_plot_x < 256)
 			{
-				//logerror("%x %x\n", pixel_plot_x, line);
-				if (tile_column == 0 && (x_scroll & 0x07))
-				{
-					/* when the first column hasn't completely entered in the screen, its
-					   background is filled only with color #0 of the selected palette */
-					line_buffer[pixel_plot_x] = m_current_palette[palette_selected ? 0x10 : 0x00];
-					priority_selected[pixel_plot_x] = priority_select;
-				}
-				else
-				{
-					line_buffer[pixel_plot_x] = m_current_palette[pen_selected];
-					priority_selected[pixel_plot_x] = priority_select | (pen_selected & 0x0f);
-				}
+				line_buffer[pixel_plot_x] = m_current_palette[pen_selected];
+				priority_selected[pixel_plot_x] = priority_select | (pen_selected & 0x0f);
 			}
 		}
 	}
 }
 
 
-void sega315_5124_device::select_sprites( int line )
+uint16_t sega315_5124_device::mode4_sprite_attributes_addr(uint16_t base)
+{
+	return base & (((m_reg[0x05] & 0x01) << 7) | 0x3f7f);
+}
+
+
+uint16_t sega315_5246_device::mode4_sprite_attributes_addr(uint16_t base)
+{
+	return base;
+}
+
+
+uint8_t sega315_5124_device::mode4_sprite_tile_mask(uint8_t tile_number)
 {
-	int max_sprites;
+	return tile_number & (((m_reg[0x06] & 0x03) << 6) | 0x3f);
+}
 
-	/* At this point the VDP vcount still doesn't refer the new line,
-	   because the logical start point is slightly shifted on the scanline */
-	int parse_line = line - 1;
+
+uint8_t sega315_5246_device::mode4_sprite_tile_mask(uint8_t tile_number)
+{
+	return tile_number;
+}
+
+
+void sega315_5124_device::select_sprites( int line )
+{
+	int max_sprites, sprite_index;
```

Comment:
Please declare `for` loop counters in the loop control statement.

Also, please declare `max_sprites` as close to the point where it becomes valid as possible.